src/commands/multi_block/monitor.rs

```diff
@@ -0,0 +1,227 @@
+			Pages = static_types::Pages,
+		> + Send
+		+ Sync
+		+ 'static,
+	T::Solution: Send,
+{
+	let signer = Signer::new(&config.seed_or_path)?;
+	let account_info = {
+		let addr = runtime::storage().system().account(signer.account_id());
+		client
+			.chain_api()
+			.storage()
+			.at_latest()
+			.await?
+			.fetch(&addr)
+			.await?
+			.ok_or(Error::AccountDoesNotExists)?
+	};
+
+	log::info!(target: LOG_TARGET, "Loaded account {} {{ nonce: {}, free_balance: {}, reserved_balance: {}, frozen_balance: {} }}",
+		signer,
+		account_info.nonce,
+		account_info.data.free,
+		account_info.data.reserved,
+		account_info.data.frozen,
+	);
+
+	let mut subscription = heads_subscription(client.rpc(), config.listen).await?;
+	let (_tx, mut rx) = tokio::sync::mpsc::unbounded_channel::<Error>();
+	let _submit_lock = Arc::new(Mutex::new(()));
+
+	let mut target_snapshot: TargetSnapshotPage = Default::default();
+	let mut voter_snapshot_paged: BTreeMap<u32, VoterSnapshotPage> = Default::default();
+
+	let n_pages = static_types::Pages::get();
+	let mut last_round_submitted = None;
+
+	loop {
+		let at = tokio::select! {
+			maybe_rp = subscription.next() => {
+				match maybe_rp {
+					Some(Ok(r)) => r,
+					Some(Err(e)) => {
+						log::error!(target: LOG_TARGET, "subscription failed to decode Header {:?}, this is bug please file an issue", e);
+						return Err(e.into());
+					}
+					// The subscription was dropped, should only happen if:
+					//	- the connection was closed.
+					//	- the subscription could not keep up with the server.
+					None => {
+						log::warn!(target: LOG_TARGET, "subscription to `{:?}` terminated. Retrying..", config.listen);
+						subscription = heads_subscription(client.rpc(), config.listen).await?;
+						continue
+					}
+				}
+			},
+			maybe_err = rx.recv() => {
+				match maybe_err {
+					Some(err) => return Err(err),
+					None => unreachable!("at least one sender kept in the main loop should always return Some; qed"),
+				}
+			}
+		};
+
+		let storage = helpers::storage_at(Some(at.hash()), client.chain_api()).await?;
+		let storage2 = storage.clone();
+		let phase = storage
+			.fetch_or_default(&runtime::storage().multi_block().current_phase())
+			.await?;
+		let round = storage.fetch_or_default(&runtime::storage().multi_block().round()).await?;
+		let desired_targets = storage
+			.fetch(&runtime::storage().multi_block().desired_targets())
+			.await
+			.unwrap()
+			.unwrap_or(0);
+
+		log::trace!(target: LOG_TARGET, "Processing block={} round={}, phase={:?}", at.number, round, phase);
+
+		match phase.0 {
+			Phase::Snapshot(page) => {
+				if page == n_pages - 1 {
+					match epm::target_snapshot(page, &storage).await {
+						Ok(snapshot) => {
+							target_snapshot = snapshot;
+						},
+						Err(err) => {
+							log::error!("fetch snapshot err: {:?}", err);
+						},
+					};
+
+					// TODO(niklasad1): The voter snapshot contains duplicates which seems wrong
```

Comment:
@kianenigma how to handle this properly?

I got duplicate votes in page 3 and page 2 and for now I just removed the votes in page 3 and just insert an empty vec there to make to `Miner::mine_solution` happy. Is it related to staking-playground or how to handle it properly?